[WlanThermo] Add support for new ESP32-powered devices [V3.x]

CSchlipp opened the pull request:
<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<!-- TITLE -->
**This PR adds support for the new Nano V3, Mini V2 (with ESP32-Upgrade), and Link V1 devices.**

<!--
Please provide a PR summary in the *Title* above, according to the following schema:
- If related to one specific add-on: Mention the add-on shortname in square brackets
  e.g. "[exec]", "[netatmo]" or "[tesla]"
- If the PR is work in progress: Add "[WIP]"
- Give a short meaningful description in imperative mood
  e.g. "Add support for device XYZ" or "Fix wrongly handled exception"
  for a new add-on/binding: "Initial contribution"
Examples:
- "[homematic] Improve communication with weak signal devices"
- "[timemachine][WIP] Initial contribution"
- "Update contribution guidelines on new signing rules"
-->

<!-- DESCRIPTION -->
The API of these hardware versions has changed compared to the previous devices, but is the same for the whole new platform (thus named "esp32").  This commit follows the same approach as the initial contribution and contains auto-generated classes for parsing the new JSON output.

The thing description xml has been spit up to different files to ease the future development.
No further changes to existing things, thus backwards compatible.

Configuration of different thing types has been merged to avoid duplicate code.

~Backport of changes to 2.5.x branch via PR #9580 (same changes)~ 

<!--
Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work
-->

<!-- TESTING -->


<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It is a good practice to add a URL to your built JAR in this Pull Request description,
so it is easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

fwolter reviewed (changes requested):
Your code doesn't build, because there are some checkstyle errors. You could take a look at `target/code-analysis/report.html`.

CSchlipp commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,222 @@
+    private final ScheduledExecutorService scheduler = ThreadPoolManager
+            .getScheduledPool(WlanThermoBindingConstants.WLANTHERMO_THREAD_POOL);
+    private final Gson gson = new Gson();
+    private @Nullable Data data = new Data();
+    private @Nullable Settings settings = new Settings();
```
TBH I don't know why this came up as an error in the V3 branch when compiling locally... 
Previous versions did not require the annotation, even the current auto-migrated binding is obviously compiling fine when run on Jenkins?

fwolter replied:
I see, Gson requires it. Then, you need to handle the null case, i.e. fix the compiler warnings about nulls.

CSchlipp replied:
Done, remaining warnings are false positives

CSchlipp commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,222 @@
+    private WlanThermoEsp32CommandHandler wlanThermoEsp32CommandHandler = new WlanThermoEsp32CommandHandler();
+    private final HttpClient httpClient;
+    private @Nullable ScheduledFuture<?> pollingScheduler;
+    private final ScheduledExecutorService scheduler = ThreadPoolManager
+            .getScheduledPool(WlanThermoBindingConstants.WLANTHERMO_THREAD_POOL);
```
Ref. discussion in initial [PR #8209](https://github.com/openhab/openhab-addons/pull/8209#discussion_r480961607), the scheduler from the BaseThingHandler constantly errored with RejectedExecutionExceptions. Wasn't able to track this down until now and went the same way as suggested for the Homematic binding in [PR #1078](https://github.com/openhab/openhab-addons/pull/1078)

fwolter replied:
AFAIK, these happen if the scheduler is shut-down. This should only happen if OH is shut-down. Can you provide a stack trace?

CSchlipp replied:
I don't have any stack traces for the scheduler issue anymore, but will try to reproduce this issue again during my local testing. 
Is this a blocker for merging or can it be tackled later on?

fwolter replied:
The RejectedExecution exception is only a symptom of a larger issue. We should at least identify the root cause before merging it.

CSchlipp replied:
Okay, I'll do some local tests to check if I can still reproduce this error in OH V3.x.

CSchlipp replied:
Stupid me - I think the error was a combination of the default scheduler shutting down and me not correctly cancelling all scheduled futures when disposing the things.
I was missing an assignment during the initialization and therefore has some orphaned tasks still running. This is fixed now, it's either the _checkConnection_ or the _update_ task running, but never more than one at the same time. 
I couldn't reproduce the error anymore, nor is the thing handler anymore complaining about updated to the thing after it got disposed. Pretty confident that this solved the issue!

fwolter replied:
Sounds good!

CSchlipp commented on bundles/org.openhab.binding.wlanthermo/src/main/resources/OH-INF/thing/channel-types.xml:
```diff
@@ -0,0 +1,261 @@
+		<state>
+			<options>
+				<option value="niagara">Niagara</option>
+				<option value="rosa">Rosa</option>
```
Color names are given by WlanThermo implementation

CSchlipp commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/data/Channel.java:
```diff
@@ -0,0 +1,125 @@
+ *
+ * @author Christian Schlipp - Initial contribution
+ */
+public class Channel {
```
Good to know, done!

CSchlipp commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,222 @@
+        data.getChannel().forEach(c -> {
+            try {
+                String json = gson.toJson(c);
+                logger.debug("Pushing: {}", json);
```
This is usually the most crucial part in case sth goes wrong. 
I'd prefer to leave it with a dedicated debug log so that users can easily include it in in bug reports.

fwolter replied:
Is there any chance to log this only if something goes wrong?

CSchlipp replied:
Payload is now only logged if response code != 200

CSchlipp commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32CommandHandler.java:
```diff
@@ -0,0 +1,272 @@
+                        break;
+                    case CHANNEL_MIN:
+                        if (command instanceof QuantityType) {
+                            channel.get(channelId).setMin(((QuantityType) command).doubleValue());
```
Changed to wildcard

CSchlipp commented:
Thanks for the review!
All required changes, except for the scheduler, are included with the last commits. Some of the improvements have also been ported to the other supported hardware versions.

Local builds do not show any blocking checkstyle errors, let's wait for the Jenkins build...

CSchlipp commented:
License Headers were missing.
Why isn't this reported during local mvn install?

CSchlipp commented:
The build failed during the Feed-Binding Integration Tests - seemingly unrelated to this binding or the changes within this PR.
Could it be related to #9610?

fwolter reviewed (changes requested):
There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32CommandHandler.java:
```diff
@@ -0,0 +1,298 @@
+        List<Channel> channel = data.getChannel();
+        if (channelUID.getId().startsWith("channel")) {
+            int channelId = Integer.parseInt(groupId.substring("channel".length())) - 1;
+            if (channel.size() > 0 && channelId < channel.size()) {
+                if (CHANNEL_ALARM_OPENHAB.equals(channelUID.getIdWithoutGroup())) {
+                    if (channel.get(channelId).getTemp() != 999) {
```
Using the `&&` operator would reduce the amount of nested indentation going on here.

CSchlipp replied:
Agree, but it wouldn't make the code clearer or better to understand. I tried to structure the different functions in the command handler all in the same way, so one can easily figure out where e.g. new Triggers have to be added.

In the end, the compiler is optimizing it anyways, so I'd plead for leaving it more readable.

cpmeister replied:
Fair enough.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32CommandHandler.java:
```diff
@@ -0,0 +1,298 @@
+                        break;
+                }
+            }
+        } else if (channelUID.getId().equals("pit1")) {
```
These channel ids should be constants specified in your WlanThermoBindingConstants file.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,222 @@
+    private WlanThermoEsp32CommandHandler wlanThermoEsp32CommandHandler = new WlanThermoEsp32CommandHandler();
+    private final HttpClient httpClient;
+    private @Nullable ScheduledFuture<?> pollingScheduler;
+    private final Gson gson = new Gson();
```
Any reason this couldn't be static final?

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32CommandHandler.java:
```diff
@@ -0,0 +1,298 @@
+ */
+public class WlanThermoEsp32CommandHandler {
+
+    public State getState(ChannelUID channelUID, Data data, Settings settings) {
```
This class is stateless, so I see no reason you couldn't make all of it's method static.
```suggestion
    public static State getState(ChannelUID channelUID, Data data, Settings settings) {
```

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,222 @@
+    private final Logger logger = LoggerFactory.getLogger(WlanThermoEsp32Handler.class);
+
+    private WlanThermoExtendedConfiguration config = new WlanThermoExtendedConfiguration();
+    private WlanThermoEsp32CommandHandler wlanThermoEsp32CommandHandler = new WlanThermoEsp32CommandHandler();
```
This declaration can go away once you make all the methods static.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,222 @@
+                authStore.addAuthentication(new DigestAuthentication(config.getUri(), Authentication.ANY_REALM,
+                        config.getUsername(), config.getPassword()));
+            }
+            pollingScheduler = scheduler.schedule(this::checkConnection, config.getPollingInterval(), TimeUnit.SECONDS);
```
Can you merge the `checkConnection` and `update` methods into a single method so that you only require a single period task to be set during `initialize`? Things can go wrong if you are assigning the `pollingScheduler` field in other threads outside of the `initialize` and `dispose` methods.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,222 @@
+                int status = httpClient.POST(uri).content(new StringContentProvider(json), "application/json")
+                        .timeout(5, TimeUnit.SECONDS).send().getStatus();
+                if (status == 401) {
+                    updateStatus(ThingStatus.ONLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            "No or wrong login credentials provided. Please configure username/password for write access to WlanThermo!");
```
A communication error should be an offline status.
```suggestion
                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
                            "No or wrong login credentials provided. Please configure username/password for write access to WlanThermo!");
```

Also if there is a credential problem here, why would you continue to send post requests for the other channels? That would just spam the logs with a bunch of failed auth attempts.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,222 @@
+                    updateStatus(ThingStatus.ONLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            "No or wrong login credentials provided. Please configure username/password for write access to WlanThermo!");
+                } else if (status != 200) {
+                    updateStatus(ThingStatus.ONLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Failed to update channel "
```
```suggestion
                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Failed to update channel "
```

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,222 @@
+                    updateStatus(ThingStatus.ONLINE);
+                }
+            } catch (InterruptedException | TimeoutException | ExecutionException | URISyntaxException e) {
+                updateStatus(ThingStatus.ONLINE, ThingStatusDetail.COMMUNICATION_ERROR,
```
```suggestion
                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
```

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/mini/WlanThermoMiniCommandHandler.java:
```diff
@@ -33,7 +36,21 @@
+        }
+
+        Unit<Temperature> unit;
+        if (app.getTempUnit().equals("fahrenheit")) {
```
So the mini uses "fahrenheit" instead of "F"?

CSchlipp replied:
yes, the API implementations unfortunately differ between the devices.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/resources/OH-INF/thing/channel-types.xml:
```diff
@@ -0,0 +1,260 @@
+	<channel-type id="rssi" advanced="true">
+		<item-type>Number:Power</item-type>
+		<label>RSSI in dBm</label>
+		<category>Text</category>
```
Why is this category `Text`?

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/resources/OH-INF/thing/channel-group-types-esp32.xml:
```diff
@@ -0,0 +1,51 @@
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+	<!-- System Group ESP32 -->
+	<channel-group-type id="cg_system_esp32">
```
Since the group types are each meant for a specific thing type, why not define the group types in the thing type files instead?

CSchlipp replied:
I've split this up to have a better overview on the different things, channel-groups and channels of the hardware versions.
Does it have any benefit of putting it in the same file?

cpmeister replied:
The only benefit is it is slightly better organized and thus requires less jumping around files to fully understand the thing xml configurations. But changing it is entirely up to you. Feel free to keep it as it is.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,222 @@
+        } else {
+            if (wlanThermoEsp32CommandHandler.setState(channelUID, command, data)) {
+                logger.debug("Data updated, pushing changes");
+                push();
```
Since `handleCommand` is called from UI threads you shouldn't make them execute potentially long running tasks like http requests. I suggest calling `push` asynchronously to prevent the UI from becoming unresponsive.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,222 @@
+                } else {
+                    updateStatus(ThingStatus.ONLINE);
+                }
+            } catch (InterruptedException | TimeoutException | ExecutionException | URISyntaxException e) {
```
An you exit the push method as soon as possible if you catch an InterruptedException since that means that openhab is trying to shutdown. You should not delay shutdown if at all possible.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,222 @@
+        if (data == null) {
+            return;
+        }
+        data.getChannel().forEach(c -> {
```
To exit early you will need to change this to regular for-each loop.

CSchlipp commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32CommandHandler.java:
```diff
@@ -0,0 +1,298 @@
+                        state = new StringType(channel.get(channelId).getName());
+                        break;
+                    case CHANNEL_TYP:
+                        state = new StringType(settings.getSensors().get(channel.get(channelId).getTyp()).getName());
```
No, this is referring to the "typ" field of the class "Channel", not to the "type" field of class "Sensor".
The naming is a bit unfortunate, but generated from the API of the device.

CSchlipp commented:
Thanks again for the review!
Checkstyle warnings should be gone now and most of the findings are included, others commented.
Ran manual tests against different devices, everything seems to be fine so far.

Jenkins build seems to be stuck with some older commit. Is it just displayed wrong in GitHub or can you retrigger it?

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32CommandHandler.java:
```diff
@@ -0,0 +1,280 @@
+                        } else if (channel.getTemp() < channel.getMin()) {
+                            return TRIGGER_ALARM_MIN;
+                        }
+                    }
+                }
```
I'd think that you would return something here just to avoid throwing a WlanThermoUnknownChannelException since it is clear at this point in the code that the channel is actually known.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,248 @@
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Failed to update channel " + c.getName() + " on device: " + e.getMessage());
+            } catch (InterruptedException e) {
+                logger.debug("Push interrupted. {}", e.getMessage());
+            }
```
You should return at this point to that you don't continue to query the other channels.
```suggestion
                logger.debug("Push interrupted. {}", e.getMessage());
                return;
            }
```

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/mini/WlanThermoMiniCommandHandler.java:
```diff
@@ -0,0 +1,171 @@
+                        } else if (data.getTemp() < data.getTempMin()) {
+                            return TRIGGER_ALARM_MIN;
+                        }
+                    }
+                }
```
see earlier comment.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/nano/WlanThermoNanoV1Handler.java:
```diff
@@ -0,0 +1,224 @@
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Failed to update channel " + c.getName() + " on device: " + e.getMessage());
+            } catch (InterruptedException e) {
+                logger.debug("Push interrupted. {}", e.getMessage());
+            }
```
```suggestion
                logger.debug("Push interrupted. {}", e.getMessage());
                return;
            }
```

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/nano/WlanThermoNanoV1CommandHandler.java:
```diff
@@ -0,0 +1,266 @@
+                case SYSTEM_CHARGE:
+                    return OnOffType.from(system.getCharge());
+                case SYSTEM_RSSI_SIGNALSTRENGTH:
+                    int dbm = system.getRssi();
+                    if (dbm >= -80) {
+                        return SIGNAL_STRENGTH_4;
+                    } else if (dbm >= -95) {
+                        return SIGNAL_STRENGTH_3;
+                    } else if (dbm >= -105) {
+                        return SIGNAL_STRENGTH_2;
+                    } else {
+                        return SIGNAL_STRENGTH_1;
+                    }
```
Where did you get these numbers from? To my knowledge a dbm of -95 is pretty much the noise floor for most wireless applications.

CSchlipp replied:
As I couldn't find any generic conversion function, I've checked the source code of the WlanThermo device and took the same numbers. 
Benefit: The signal indicator now gives the same number of bars as shown on the display of the device.

cpmeister replied:
Well if it matches what is shown on the device then I guess that is the correct implementation. Keep it as is then.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/nano/WlanThermoNanoV1CommandHandler.java:
```diff
@@ -0,0 +1,266 @@
+                Pm pm = data.getPitmaster().getPm().get(0);
+                switch (channelUID.getIdWithoutGroup()) {
+                    case CHANNEL_PITMASTER_CHANNEL_ID:
+                        pm.setChannel(((QuantityType<?>) command).intValue());
+                        return true;
+                    case CHANNEL_PITMASTER_PIDPROFILE:
+                        pm.setPid(((QuantityType<?>) command).intValue());
+                        return true;
+                    case CHANNEL_PITMASTER_SETPOINT:
+                        pm.setSet(((QuantityType<?>) command).toUnit(unit).doubleValue());
```
You should make sure that the units are converted to what you expect them to be. Use `QuantityType.toUnit` to handle the conversion.

CSchlipp replied:
That already included in all channels, where unit conversion might be required.
The PITMASTER_PIDPROFILE channel however only contains an integer reference to a profile, so no conversion needed.

cpmeister replied:
If PITMASTER_PIDPROFILE is just a integer reference then why does it need to be a `QuantityType`?
Same question for CHANNEL_PITMASTER_CHANNEL_ID.

CSchlipp replied:
Ah, now I understood what you meant. 
Reverted to DecimalType.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/nano/WlanThermoNanoV1CommandHandler.java:
```diff
@@ -0,0 +1,266 @@
+                        }
+                }
+            }
+        } else if (channelUID.getId().equals(CHANNEL_PITMASTER_1)) {
```
```suggestion
        } else if (channelUID.getId().startsWith(CHANNEL_PITMASTER_1)) {
```

CSchlipp replied:
That's actually a difference in the NanoV1 device, it does only provide a single Pitmaster channel.
Therefore, I can directly check for the whole id instead of checking the prefix first and then obtaining the actual channel numbers.

cpmeister replied:
But just a couple of lines down you are expecting `channelUID.getIdWithoutGroup()` to return different results? Wouldn't the `CHANNEL_PITMASTER_1` be the only result that could be returned? As the code is written now the switch statement serves no purpose.

CSchlipp replied:
True, mixed up group and ID. Fixed.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,248 @@
+                }
+
+                // Update Channels if required
+                Map<String, String> properties = editProperties();
+                Boolean pmEnabled = settings.getFeatures().getBluetooth();
+                int pmChannels = pmEnabled ? data.getPitmaster().getPm().size() : 0;
+                int tempChannels = data.getChannel().size();
+
+                // Update properties
+                properties.put(WlanThermoBindingConstants.PROPERTY_MODEL, settings.getDevice().getDevice());
+                properties.put(WlanThermoBindingConstants.PROPERTY_SERIAL, settings.getDevice().getSerial());
+                properties.put(WlanThermoBindingConstants.PROPERTY_ESP32_BT_ENABLED,
+                        settings.getFeatures().getBluetooth().toString());
+                properties.put(WlanThermoBindingConstants.PROPERTY_ESP32_PM_ENABLED, pmEnabled.toString());
+                properties.put(WlanThermoBindingConstants.PROPERTY_ESP32_TEMP_CHANNELS, String.valueOf(tempChannels));
+                properties.put(WlanThermoBindingConstants.PROPERTY_ESP32_PM_CHANNELS, String.valueOf(pmChannels));
+                updateProperties(properties);
```
Why are you changing the properties every time? Thing properties should generally be considered immutable characteristics of the device so you should only need to update it when something changes.

CSchlipp replied:
The number of temperature and pitmaster channels might change during operation as the user attaches bluetooth sensors. 
On the one hand, I do understand that properties should usually be immutable, but on the other hand I don't think that having the number of attached sensors is worth an own channel. 
I've changed the code so that only the channel number properties are updated, is that okay for you?

cpmeister replied:
It wasn't really what I had in mind but it doesn't really sound like you can avoid updating the properties. So I guess this is fine.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/nano/WlanThermoNanoV1Handler.java:
```diff
@@ -0,0 +1,229 @@
+        }
+
+        // push update for pitmaster channels
+        try {
+            String json = gson.toJson(data.getPitmaster().getPm());
+            URI uri = config.getUri("/setpitmaster");
+            int status = httpClient.POST(uri).content(new StringContentProvider(json), "application/json")
+                    .timeout(5, TimeUnit.SECONDS).send().getStatus();
+            if (status == 401) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "No or wrong login credentials provided. Please configure username/password for write access to WlanThermo!");
+            } else if (status != 200) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Failed to update pitmaster channel on device, Statuscode " + status + " on URI "
+                                + uri.toString());
+                logger.debug("Payload sent: {}", json);
+            } else {
+                updateStatus(ThingStatus.ONLINE);
+            }
+        } catch (TimeoutException | ExecutionException | URISyntaxException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Failed to update pitmaster channel on device: " + e.getMessage());
+        } catch (InterruptedException e) {
+            logger.debug("Push interrupted. {}", e.getMessage());
+        }
```
This code is nearly identical to the code right above it. Please consider refactoring this into a separate method.

CSchlipp replied:
Done

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32Handler.java:
```diff
@@ -0,0 +1,253 @@
+            }
+        }
+
+        // push update for pitmaster channels
+        try {
+            String json = gson.toJson(data.getPitmaster().getPm());
+            URI uri = config.getUri("/setpitmaster");
+            int status = httpClient.POST(uri).content(new StringContentProvider(json), "application/json")
+                    .timeout(5, TimeUnit.SECONDS).send().getStatus();
+            if (status == 401) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "No or wrong login credentials provided. Please configure username/password for write access to WlanThermo!");
+            } else if (status != 200) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Failed to update pitmaster channel on device, Statuscode " + status + " on URI "
+                                + uri.toString());
+                logger.debug("Payload sent: {}", json);
+            } else {
+                updateStatus(ThingStatus.ONLINE);
+            }
+        } catch (TimeoutException | ExecutionException | URISyntaxException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Failed to update pitmaster channel on device: " + e.getMessage());
+        } catch (InterruptedException e) {
+            logger.debug("Push interrupted. {}", e.getMessage());
+        }
```
Same suggestion here.

cpmeister reviewed (approved):
I ran out of things to complain about. Looks good to me now.

CSchlipp commented:
> I ran out of things to complain about. Looks good to me now.

Thanks for the review, appreciate it!

cpmeister commented:
Just need to get a completed jenkins build that succeeds. Did you ever address @fwolter's comment about the `target/code-analysis/report.html`?

CSchlipp commented:
> Just need to get a completed jenkins build that succeeds. Did you ever address @fwolter's comment about the `target/code-analysis/report.html`?

I couldn't find any checkstyle warnings in my local build anymore. 
There are some warnings about potential null pointer access due to Gson having @Nullable return objects. All of them should now be covered by a check for null before accessing the object, so this should be false positives.

CSchlipp commented:
> > Just need to get a completed jenkins build that succeeds. Did you ever address @fwolter's comment about the `target/code-analysis/report.html`?
> 
> I couldn't find any checkstyle warnings in my local build anymore.
> There are some warnings about potential null pointer access due to Gson having @nullable return objects. All of them should now be covered by a check for null before accessing the object, so this should be false positives.

Just for the sake of completeness:
The minor checkstyle findings for MemberNameCheck on the autogenerated DTO  are the same as before this PR, but checkstyle step passes. 
As this class is auto-generated, I'd prefer not to have manual changes in it.

Last report:

class file | high | medium | low | total
-- | -- | -- | -- | --
org.openhab.binding.wlanthermo.internal.api.mini.dto.builtin.Channel.java | 0 | 0 | 10 | 10
  | 0 | 0 | 10 | 10

tool |	priority |	line |	category |	rule |	message
-- | -- | -- | -- | -- | --
checkstyle |	3 |	28 |	style |	MemberNameCheck |	Der Bezeichner '_0' entspricht nicht dem Muster '^[a-z][a-zA-Z0-9]*$'.
checkstyle |	3 |	31 |	style	| MemberNameCheck |	Der Bezeichner '_1' entspricht nicht dem Muster '^[a-z][a-zA-Z0-9]*$'.
checkstyle |	3 |	34 |	style |	MemberNameCheck |	Der Bezeichner '_2' entspricht nicht dem Muster '^[a-z][a-zA-Z0-9]*$'.
checkstyle |	3 |	37 |	style |	MemberNameCheck |	Der Bezeichner '_3' entspricht nicht dem Muster '^[a-z][a-zA-Z0-9]*$'.
checkstyle |	3 |	40 |	style |	MemberNameCheck |	Der Bezeichner '_4' entspricht nicht dem Muster '^[a-z][a-zA-Z0-9]*$'.
checkstyle |	3 |	43 |	style |	MemberNameCheck |	Der Bezeichner '_5' entspricht nicht dem Muster '^[a-z][a-zA-Z0-9]*$'.
checkstyle |	3 |	46 |	style |	MemberNameCheck |	Der Bezeichner '_6' entspricht nicht dem Muster '^[a-z][a-zA-Z0-9]*$'.
checkstyle |	3 |	49 |	style |	MemberNameCheck |	Der Bezeichner '_7' entspricht nicht dem Muster '^[a-z][a-zA-Z0-9]*$'.
checkstyle |	3 |	52 |	style |	MemberNameCheck |	Der Bezeichner '_8' entspricht nicht dem Muster '^[a-z][a-zA-Z0-9]*$'.
checkstyle |	3 |	55 |	style |	MemberNameCheck |	Der Bezeichner '_9' entspricht nicht dem Muster '^[a-z][a-zA-Z0-9]*$'.

CSchlipp commented:
Argh, missed the last spotless:apply commit. Would you label it for rebuild again, please? 
Sorry for the inconvenience caused...

cpmeister commented:
> Argh, missed the last spotless:apply commit. Would you label it for rebuild again, please?
> Sorry for the inconvenience caused...

No problem.

cpmeister commented:
> > Just need to get a completed jenkins build that succeeds. Did you ever address @fwolter's comment about the `target/code-analysis/report.html`?
> 
> I couldn't find any checkstyle warnings in my local build anymore.
> There are some warnings about potential null pointer access due to Gson having @nullable return objects. All of them should now be covered by a check for null before accessing the object, so this should be false positives.

Actually, the potential null pointer access warnings I see there would be caused by using nullable fields directly instead of caching the field into a local variable first. The null checker assumes that any nullable field can become null at any time due to concurrency. And yes, I understand that it is often a wrong assumption but it is also the safest assumption.
So please make an effort to address those null pointer warnings.

CSchlipp commented:
> Actually, the potential null pointer access warnings I see there would be caused by using nullable fields directly instead of caching the field into a local variable first. The null checker assumes that any nullable field can become null at any time due to concurrency. And yes, I understand that it is often a wrong assumption but it is also the safest assumption.
> So please make an effort to address those null pointer warnings.

Guess I've found a quite pretty solution for the null pointer warnings.
Additionally, I've added unit tests for all models and moved a lot of duplicated code & logic to parent classes.

Would you please trigger a rebuild? Thanks!

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/WlanThermoHandler.java:
```diff
@@ -0,0 +1,190 @@
+            for (Channel channel : thing.getChannels()) {
+                updateState(channel.getUID(), UnDefType.UNDEF);
+            }
+            throw new WlanThermoInputException();
```
Always pass the cause exceptions into new exceptions you throw.
Please add such a constructor is one doesn't already exist.
```suggestion
            throw new WlanThermoInputException(e);
```

Also, I don't think that a WlanThermoInputException is the correct thing to throw here if you get a TimeoutException. The WlanThermoInputException message implies incorrect data but a Timeout is not related to that at all, so the exception message would be misleading. Consider having the `doGet` method throw WlanThermoException instead so you can throw broader types of exceptions.

CSchlipp replied:
A Timeout would however result in invalid input data for the binding.
Changed it anyways.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/WlanThermoHandler.java:
```diff
@@ -0,0 +1,190 @@
+    protected final HttpClient httpClient;
+    protected final Logger logger = LoggerFactory.getLogger(WlanThermoHandler.class);
+    protected final Gson gson = new Gson();
+    @Nullable
+    protected ScheduledFuture<?> pollingScheduler;
```
```suggestion
    protected @Nullable ScheduledFuture<?> pollingScheduler;
```

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/test/java/org/openhab/binding/wlanthermo/internal/api/esp32/WlanThermoEsp32CommandHandlerTest.java:
```diff
@@ -0,0 +1,601 @@
+    private static final ThingUID THING_UID = new ThingUID("wlanthermo", "esp32", "test");
+
+    //@formatter:off
+    private static final String DATA_INPUT_JSON = "{\n" +
```
Please put test data in their own files and then load them in runtime. You can do this by loading the resources from the ClassLoader.

cpmeister commented on bundles/org.openhab.binding.wlanthermo/src/main/java/org/openhab/binding/wlanthermo/internal/api/nano/WlanThermoNanoV1CommandHandler.java:
```diff
@@ -0,0 +1,275 @@
+                        if (command instanceof StringType) {
+                            channel.setName(command.toFullString());
+                            return true;
+                        }
+                    case CHANNEL_MIN:
```
Is this fall through intentional?

CSchlipp replied:
No, this bug was introduced by the change of the return handling. Thanks, fixed.

cpmeister reviewed (approved):
LGTM

cpmeister commented:
The build failed, looks like you need to run spotless.

CSchlipp commented:
This is driving me nuts some day...
Sorry, fixed.

CSchlipp commented:
Thanks again!
I've updated the PR name to be more precise in the changelogs.
